src/main/java/org/apache/ibatis/builder/annotation/AbstractProviderSqlSource.java

```diff
@@ -0,0 +1,99 @@
+import org.apache.ibatis.session.Configuration;
+
+import java.lang.reflect.Method;
+
+/**
+ * @author Clinton Begin
+ * @author Kazuki Shimizu
+ * @author liuzh
+ * @since 3.4.6
+ */
+public abstract class AbstractProviderSqlSource implements SqlSource {
+
+  protected final Configuration configuration;
+  protected final SqlSourceBuilder sqlSourceParser;
+  protected final Class<?> providerType;
+  protected Method providerMethod;
+  protected String[] providerMethodArgumentNames;
+  protected Class<?>[] providerMethodParameterTypes;
+  protected ProviderContext providerContext;
+  protected Integer providerContextIndex;
+
+  /**
+   * @since 3.4.6
+   */
+  public AbstractProviderSqlSource(Configuration configuration, Object provider, Class<?> mapperType, Method mapperMethod) {
+    String providerMethodName;
+    try {
+      this.configuration = configuration;
+      this.sqlSourceParser = new SqlSourceBuilder(configuration);
+      this.providerType = (Class<?>) provider.getClass().getMethod("type").invoke(provider);
+      providerMethodName = (String) provider.getClass().getMethod("method").invoke(provider);
+
+      for (Method m : this.providerType.getMethods()) {
+        if (providerMethodName.equals(m.getName())) {
+          if (m.getReturnType() == String.class) {
+            if (providerMethod != null){
+              throw new BuilderException("Error creating SqlSource for SqlProvider. Method '"
+                      + providerMethodName + "' is found multiple in SqlProvider '" + this.providerType.getName()
+                      + "'. Sql provider method can not overload.");
+            }
+            this.providerMethod = m;
+            this.providerMethodArgumentNames = new ParamNameResolver(configuration, m).getNames();
+            this.providerMethodParameterTypes = m.getParameterTypes();
+          }
+        }
+      }
+    } catch (BuilderException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new BuilderException("Error creating SqlSource for SqlProvider.  Cause: " + e, e);
+    }
+    if (this.providerMethod == null) {
+      throw new BuilderException("Error creating SqlSource for SqlProvider. Method '"
+          + providerMethodName + "' not found in SqlProvider '" + this.providerType.getName() + "'.");
+    }
+    for (int i = 0; i< this.providerMethodParameterTypes.length; i++) {
+      Class<?> parameterType = this.providerMethodParameterTypes[i];
+      if (parameterType == ProviderContext.class) {
+        if (this.providerContext != null){
+          throw new BuilderException("Error creating SqlSource for SqlProvider. ProviderContext found multiple in SqlProvider method ("
+              + this.providerType.getName() + "." + providerMethod.getName()
+              + "). ProviderContext can not define multiple in SqlProvider method argument.");
+        }
+        this.providerContext = new ProviderContext(mapperType, mapperMethod);
+        this.providerContextIndex = i;
```

Comment:
this line of code is guaranteed to be` this.providerContextIndex = this.providerMethodParamterTypes.length -1`
You can remove it from the `for` loop. And, the creation of a new instance of `ProviderContext` each and every time is just as redundant but creates object instances unnecessarily